### ISSUE TITLE
Support listener pod annotations

### ISSUE THREAD

**Issue by int128** [actions/actions-runner-controller#2881]
### What would you like added?
It would be nice if we can set a listener pod spec such as `.metadata.annotations`.

### Why is this needed?
We'd like to monitor the metrics of listener pods by Datadog. It requires the pod annotation like:

```yaml
apiVersion: v1
kind: Pod
metdata:
  annotations:
    ad.datadoghq.com/autoscaler.checks: |
      {
        "openmetrics": {
          "init_config": {},
          "instances": [
            {
              "openmetrics_endpoint": "http://%%host%%:8080/metrics",
              "namespace": "arc",
              "metrics": ["gha_*"]
            }
          ]
        }
      }
```

Currently, there is no way to set annotations to a listener pod.

### Additional context
I notice the change of https://github.com/actions/actions-runner-controller/pull/2758. I hope this issue will be also resolved.

**Comment by nikola-jokic** [comment]
Hey @int128,

This should be resolved with the ability to extend the listener spec. That is the reason why we chose PodTemplateSpec, so that user can modify the spec, as well as provide additional annotations and labels :relaxed:

**Comment by int128** [comment]
@nikola-jokic Thank you for your effort! I tried the new release but did not work for our use-case.

## Attempt 1
I changed our runner as follows:

```yaml
apiVersion: actions.github.com/v1alpha1
kind: AutoscalingRunnerSet
spec:
  listenerTemplate:
    metadata:
      annotations:
        ad.datadoghq.com/manager.checks: |
          {
            "openmetrics": { ...omit... }
          }
```

but got the following log in the controller:

> 2023-09-21T00:24:38Z	ERROR	Reconciler error	{"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "AutoscalingRunnerSet": {"name":"example","namespace":"arc-runners"}, "namespace": "arc-runners", "name": "example", "reconcileID": "6d924315-b56f-441f-936b-c7783d5786a4", "error": "AutoscalingListener.actions.github.com \"example-754b578d-listener\" is invalid: spec.template.spec.containers: Required value"}

I think this is a limitation of PodTemplateSpec, so then tried to put an empty field.

## Attempt 2
I changed our runner as follows:

```yaml
apiVersion: actions.github.com/v1alpha1
kind: AutoscalingRunnerSet
spec:
  listenerTemplate:
    metadata:
      annotations:
        ad.datadoghq.com/manager.checks: |
          {
            "openmetrics": { ...omit... }
          }
  spec:
    containers: []
```

but the controller becomes CrashLoopBackoff due to the error:

```
2023-09-21T00:43:02Z	INFO	AutoscalingListener	Creating a listener pod	{"autoscalinglistener": "arc-systems/example-754b578d-listener"}
2023-09-21T00:43:02Z	INFO	Observed a panic in reconciler: assignment to entry in nil map	{"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "AutoscalingListener": {"name":"example-754b578d-listener","namespace":"arc-systems"}, "namespace": "arc-systems", "name": "example-754b578d-listener", "reconcileID": "c553f2e1-ab06-4231-b9c6-7fdbf0760ac7"}
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 278 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x18570a0, 0x1d111c0})
	runtime/panic.go:884 +0x213
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.mergeListenerPodWithTemplate(0xc00514a000, 0xc000623500)
	github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:323 +0x10c
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*resourceBuilder).newScaleSetListenerPod(0x1518e4b?, 0xc003fd6000, 0xc00309f040, 0xc003278280, 0xc003fd5938, {0x0, 0x0, 0x1ab8562?})
	github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:313 +0x13cd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).createListenerPod(0xc0000adaa0, {0x1d2a760, 0xc003d81980}, 0x0?, 0xc003fd6000, 0xc004042540?, 0xc003fd5aa0?, {{0x1d2cb10, 0xc003d819b0}, 0x0})
	github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:406 +0x8bd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).Reconcile(0xc0000adaa0, {0x1d2a760, 0xc003d81980}, {{{0xc0006a1280, 0xb}, {0xc0006a75e0, 0x1f}}})
	github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:243 +0x2110
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1d2a760?, {0x1d2a760?, 0xc003d81980?}, {{{0xc0006a1280?, 0x17cd5a0?}, {0xc0006a75e0?, 0x1d18948?}}})
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000619ea0, {0x1d2a6b8, 0xc0000af590}, {0x18d1d40?, 0xc003bb43a0?})
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:323 +0x377
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000619ea0, {0x1d2a6b8, 0xc0000af590})
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:231 +0x587
```

I have created a fix as https://github.com/actions/actions-runner-controller/pull/2913. Please pick it if you like.

**Comment by nikola-jokic** [comment]
Hey @int128,

You are right, sorry I introduced a bug
I extended your PR I hope you don't mind :relaxed:

**Comment by int128** [comment]
I set the pod annotations and the controller works perfectly.
Thank you very much! @nikola-jokic 

fyi: Here is a snippet to collect the listener metrics by Datadog.

```yaml
apiVersion: actions.github.com/v1alpha1
kind: AutoscalingRunnerSet
spec:
  listenerTemplate:
    metadata:
      annotations:
        ad.datadoghq.com/autoscaler.checks: |
          {
            "openmetrics": {
              "init_config": {},
              "instances": [
                {
                  "openmetrics_endpoint": "http://%%host%%:8080/metrics",
                  "namespace": "arc",
                  "metrics": ["gha_job_.*"],
                  "histogram_buckets_as_distributions": true
                }
              ]
            }
          }
  spec:
    containers: []
```

**Comment by samchoraria** [comment]
Hi @int128 & @nikola-jokic I have a question about the `listenerTemplate` configuration.

I can apply the following annotations to the `manager` via Helm chart `arc/gha-runner-scale-set-controller v0.6.1` values.
```
podAnnotations:
  ad.datadoghq.com/manager.checks: |
    {
      "openmetrics": {
        "init_config": {},
        "instances": [
          {
            "openmetrics_endpoint": "http://%%host%%:8080/metrics",
            "namespace": "arc_manager",
            "metrics": [".*"]
          }
        ]
      }
    }
```
And I see metrics arrive in Datadog under namespace `arc_manager.*`

Using config [snippet](https://github.com/actions/actions-runner-controller/issues/2881#issuecomment-1746446923) from @int128 I’m unable to get `listener` metrics into Datadog _also_.
I applied the config change to a running cluster via Helm chart `arc/gha-runner-scale-set v0.6.1` values:
```
listenerTemplate:
  metadata:
    annotations:
      ad.datadoghq.com/autoscaler.checks: |
        {
          "openmetrics": {
            "init_config": {},
            "instances": [
              {
                "openmetrics_endpoint": "http://%%host%%:8080/metrics",
                "namespace": "arc_autoscaler",
                "metrics": [".*"]          
              }
            ]
          }
        }
  spec:
    containers: []
```
@int128, did you use the helm chart & values or another approach to apply changes to `listenerTemplate` in the `AutoscalingRunnerSet`?

I also tried using annotation key `ad.datadoghq.com/listener.checks:` with `spec.containers[{name: 'listener'}]` which is consistent with Datadog [documentation](https://docs.datadoghq.com/containers/kubernetes/prometheus/?tab=kubernetesadv2)

**UPDATE:**
The problem I was hitting was that `listenerTemplate` annotations were NOT applied to the listener pod when making changes via Helm.  It was necessary for me to delete the existing CRDs (e.g. `AutoscalingListener` in particular) - I was then able to redeploy both Helm charts at `v0.7.0` and see the annotations get applied to the listener pod.
